README
README FILE FOR PERL MODULE -- DBD::AnyData

WHAT THE HECK IS IT?

  The DBD::AnyData module provides a DBI (Perl Database Interface)
  and SQL (Structured Query Language) interface to data in many 
  formats and from many sources.

  There are actually two modules DBD::AnyData and AnyData.  The AnyData
  module provides most of the same features as DBD::AnyData
  through a tied hash interface which does not require or support
  DBI and SQL.  

  Both modules allow accessing and modifying data in over a dozen
  formats.  If the DBI/SQL approach is the one you want, read on.
  Otherwise see the documentation for AnyData.pm

  Currently supported formats include general format flatfiles (CSV,
  Fixed Length, Tab or Pipe "delimited", etc.), specific formats 
  (passwd files, web logs, etc.), a variety of other kinds of 
  formats (XML, Mp3, HTML tables), and, for some
  operations, any DBI accessible database. The number of supported
  formats will continue to grow rapidly since there is an open API
  making it easy for any author to create additional format parsers
  which can be plugged in to AnyData.

  Data in these various formats can come from local files, from
  remote files, or from perl data structures such as strings and
  arrays.

  Regardless of the format or source of the data, it may be accessed
  and/or modified using all standard DBI methods and a subset of SQL
  syntax.

  In addition to standard database access to files, the module also
  supports in-memory tables which allow you to create temporary
  views; to combine data from a number of sources; to quickly
  prototype database systems; and to display or save the data in any
  of the supported formats (e.g. to display data in a CSV file as an
  HTML table).  These in-memory tables can be created from any 
  combination of DBI databases or files of any format.  They may also
  be created from perl data structures which means it's possible to
  quickly prototype a database system without any file access or 
  rdbms backend.

  The module also supports converting files between any of the supported
  formats (e.g. save selected data from MySQL or Oracle to an XML file).

SOME EXAMPLES

  # SELECT DATA FROM A PASSWD FILE
  #
  $dbh->func( 'users', 'Passwd', '/etc/passwd', 'ad_catalog');
  my $sth = $dbh->prepare("SELECT username,homedir,GID FROM users');

  # INSERT A NEW ROW INTO A CSV FILE
  #
  $dbh->func( 'cars', 'CSV', 'cars.csv', 'ad_catalog');
  $dbh->do("INSERT INTO cars VALUES ('Honda','Odyssey')");

  # READ A REMOTE XML FILE AND PRINT IT AS AN HTML TABLE
  #
  print $dbh->func( 'XML', $url, 'HTMLtable', 'ad_convert');

  # CONVERT A MYSQL DATABASE INTO XML AND SAVE IT IN A NEW FILE
  #
  $dbh->func( 'DBI', $mysql_dbh, 'XML', 'data.xml', 'ad_convert');

  # CREATE AND ACCESS A VIEW CONTAINING DATA FROM 
  # AN ORACLE DATABASE AND A TAB DELIMITED FILE
  #
  $dbh->func( 'combo', 'DBI', $oracle_dbh, 'ad_import');
  $dbh->func( 'combo', 'Tab', 'data.tab', 'ad_import');
  my $sth = $dbh->prepare("SELECT * FROM combo");

  # CREATE  A TEST TABLE FROM A REFERENCE TO AN ARRAY OF ARRAYS
  #
  $dbh->func( 'test', 'ARRAY', $array_ref, 'ad_import');
  my $sth = $dbh->prepare("SELECT * FROM test WHERE foo='bar'");

  # CREATE  A TEST TABLE FROM THE DATA SECTION OF A SCRIPT
  #
  $dbh->func( 'test', 'XML', [<DATA>], 'ad_import');
  my $sth = $dbh->prepare("SELECT phrase FROM test WHERE id=2");
  __END__
  <phrases>
      <phrase id="1">Hello World!</phrase>
      <phrase id="2">Just Another Perl Hacker</phrase>
  </phrases>

WHAT ELSE DO I NEED?

  To use DBD::AnyData you will need to install these modules,
  all available from CPAN and most available from activeState.

  * DBI
  * SQL::Statement
  * DBD::File
  * AnyData
  * DBD::AnyData

  Note: DBD::File is part of the DBD::CSV distribution

  Additional modules are required for some advanced features,
  see 'perldoc DBD::AnyData'.

HOW DO I INSTALL IT?

   1.  Install all the prerequired modules first if they 
       are not already installed

   2.  Unpack the compressed files.
       (DBD-AnyData-version.tar.gz or DBD-AnyData-version.zip)

   3a. If you are not familiar with the standard Perl
       makefile method, you can simply copy the file DBD-AnyData.pm
       into your ...site/lib/DBD directory
  
   3b. If you are familiar with the standard Perl make 
       installation, just do as always (perl Makefile.PL; 
       make; make test; make install) this should also 
       work with dmake or nmake. 

WHERE CAN I GET MORE INFO?

   After installing the module, type "perldoc DBD::AnyData" at 
   the command prompt, or just read the documentation at 
   the bottom of the ...DBD/AnyData.pm file.

WHO DUNNIT?

  Jeff Zucker <jeff@vpservices.com>

Enjoy!





Changes
DBD::AnyData - DBI access to XML,CSV and other data

Revision history for Perl extension DBD::AnyData.

version 0.120 release 

* new maintainer Sven Dowideit
* fixes ...

version 0.110 release 17th July 2010

* new maintainer Jens Rehsack
* fixes to run with DBI-1.612

version 0.09 released 15 June 2007

* finally fixed the "cleared while still active" bug

version 0.08, released 18 April 2004

 * now sets $dbh->{Active} on with $dbh->connect and off
   with $dbh->disconnect; this allows it to work with
   Ima::DBI, Class::DBI and other environments that cache
   a database handle

version 0.07, released 15 March 2004

 * fixed problem with DBD::File-0.03 STORE
   (thanks to Dan Wright for bug report)

 * fixed ad_import with DBI and keep_connection flag to operate according
   to the docs (when set, allows you to keep a connection)
   (thanks to Dan Wright for bug report)

 * fixed func() calls so they can optionally be called without func()
   if you have DBI version 1.37 or higher, for example these are the same:

     $dbh->ad_dump('SELECT * FROM old');
     $dbh->func('SELECT * FROM old','ad_dump');

   this also applies to ad_import(), ad_convert(), ad_catalog(),
   ad_export() and ad_clear()

version 0.06, released 25 August 2003

 prepare()   : added caching of parser
 Makefile.PL : added prereqs, (thx Scott Godin and Intrepid)
 test.pl     : changes $sth->{NAME} to $th->{NAME_lc}, omitted XML & HTML tests

version 0.05, released 17 July, 2001

 There are major changes in the way the tied hash interface
 does deletions and exporting and in the XML and Weblog format
 parsers.  If you use the tied hash interface or XML or Weblog
 (in either interface), please update to this version.

 The Big Stuff:

   * XML now accepts user-supplied tag-to-column mappings.

   * Import now supports cross and outer joins by importing more than
     one table into the same in-memory table.

 Many minor fixes:

 * fixed adExport to allow flags as documented
   (thx Matthew Wickline)

 * greatly improved and speeded up deletions from tied hashes
   and documented how they work (thx *alot* Matthew Wickline)

 * fixed weblog to handle embedded double quotes
   (thx Bob O'Neill)

 * added documentation to DBD::AnyData on cross joins and full
   outer joins as part of the import statement

 * fixed tests so they skip XML and HTMLtable if the
   required extra modules (e.g. XML::Twig) aren't available
   (thx Malcolm Cook)

 * fixed XML to be able to work with XML::Twig 3.00 as well
   as earlier versions (meant to do that earlier)

 * fixed README to mention that DBD::AnyData replaces
   DBD::RAM (thx Mark Whittiker)

 * fixed weblog README to use current method and column names
   (thx Bob O'Neill)

 * fixed HTMLtable so that it can be used to export even if
   HTML::TableExtract and HTML::Parser aren't installed
   (thx Matthew Wickline)

 * fixed bad version number on DBD::AnyData - it reported 0.03
   instead of 0.04 (hey I caught this one all by myself)

 * added version numbers to all of the submodules and standardized
   the docs

Version 0.04, 26th June, 2001

 * Beta release

Version 0.03, 17th April, 2001

 * second Alpha release

Version 0.02, 4th February, 2001

 * first Alpha release

Version 0.01, 17th January, 2001

 * pre Alpha release

Changes from DBD::RAM to DBD::AnyData

  * streamlined interface, faster operations

  * now supports new formats: HTMLtable, Passwd, Weblog, Paragraph

  * import() now supports SELECT (import only the columns and rows
    you specify) and joining multiple tables into a single table

  * a new convert() method allows one-step conversion between any
    of the supported formats

  * additional support for remote files including authentication

  * additional support for XML including greatly improved export()
    and use of the document's DTD for both importing and exporting

Gotchas in moving from DBD::RAM to DBD::AnyData

  * the func() methods now all start with ad_, thus ad_catalog instead
    of catalog



Build.PL
use strict;
use warnings;
use Module::Build;

my $builder = Module::Build->new(
    module_name       => 'DBD::AnyData',
    license           => 'perl',
    #author       => [ q{Jens Rehsack <rehsack@cpan.org>}, q{Sven Dowideit <SvenDowideit@fosiki.com>} ],
    author_from => 'lib/DBD/AnyData.pm',
    dist_version_from => 'lib/DBD/AnyData.pm',
    resources         => {
                   repository  => 'https://github.com/SvenDowideit/DBD-AnyData',
                   MailingList => 'mailto:dbi-dev@perl.org',
                   license     => 'http://dev.perl.org/licenses/',
                 },
    build_requires => { 'Test::More' => 0.90, },
    requires       => {
                  'AnyData'        => '0.07',
                  'DBI'            => '1.611_93',
                  'SQL::Statement' => '1.27_02',
                  'Params::Util'   => '1.00',
                },
    recommends => {
                    'AnyData'        => '0.10',
                    'DBI'            => '1.612',
                    'SQL::Statement' => '1.28',
                    'Params::Util'   => '1.01',
                  },
    # from doc, until AnyData has been converted
    auto_features => {
                       remote_files => {
                                         description => "Remote file access",
                                         requires    => { 'LWP' => 0, },
                                         recommends  => { 'LWP' => '5.836', },
                                       },
                       xml_access => {
                                       description => "XML access",
                                       requires    => {
                                                     'XML::Parser' => 0,
                                                     'XML::Twig'   => 0,
                                                   },
                                       recommends => {
                                                       'XML::Parser' => '2.36',
                                                       'XML::Twig'   => '3.35',
                                                     },
                                     },
                       html_tables => {
                                        description => "HTML Table acess",
                                        requires    => {
                                                      'HTML::Parser'       => 0,
                                                      'HTML::TableExtract' => 0,
                                                    },
                                        recommends => {
                                                        'HTML::Parser'       => '3.65',
                                                        'HTML::TableExtract' => '2.10',
                                                      },
                                      },
                       html_table_writing => {
                                               description => "Write HTML Tables",
                                               requires    => { 'CGI' => 0, },
                                               recommends  => { 'CGI' => '3.49', },
                                             },
                     },
    CONFLICTS => {
                   'SQL::Statement' => '1.27',
                   'DBI'            => '1.611',
                 },
    add_to_cleanup     => ['DBD-AnyData-*'],
    create_makefile_pl => 'traditional',
                                );

$builder->create_build_script();
